backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/UserService.cs

```diff
@@ -59,90 +59,121 @@ public async Task<UserDto> AddTagsAsync(List<TagDto> tags)
+        if (filter.Language is not null)
+        {
+            filteredUsers = filteredUsers.Where(u => u.Language == filter.Language);
+        }
+        if (filter.LangLevels is not null && filter.LangLevels.Any())
+        {
+            filteredUsers = filteredUsers.Where(u => filter.LangLevels.Contains(u.LanguageLevel));
+        }
+        if (filter.Topics is not null && filter.Topics.Any())
+        {
+            filteredUsers = filteredUsers.Where(u => u.Tags.Any(t => filter.Topics.Contains(t.Name)));
+        }
+        var filteredUsersList = await filteredUsers.ToListAsync();
+        return _mapper.Map<List<UserShortInfoDto>>(filteredUsersList);
+    }
+
+    public async Task<LessonDto> EnrollUserToLesson(long lessonId)
+    {
+        var userId = _authService.UserId;
+
+        var user = _context.Users.SingleOrDefault(u => u.Id == userId) ?? throw new ArgumentException($"Failed to find the user with id {userId}");
+        var lesson = _context.Lessons.SingleOrDefault(l => l.Id == lessonId) ?? throw new ArgumentException($"Failed to find the lesson with id {lessonId}");
+
+        user.Lessons.Add(lesson);
+
+        await _context.SaveChangesAsync();
+
+        void AfterMapAction(Lesson o, LessonDto dto) => dto.SubscribersCount = _context.Lessons
+            .Select(t => new { Id = t.Id, SbCount = t.Subscribers.Count })
+            .FirstOrDefault(l => l.Id == lessonId)!.SbCount;
+
+        return _mapper.Map<Lesson, LessonDto>(lesson, options => options.AfterMap(AfterMapAction));
+    }
+
+    public async Task<string> UploadProfilePhoto(IFormFile file)
+    {
+        var user = await _context.Users.FirstOrDefaultAsync(u => u.Id == _authService.UserId);
+
+        if (user == null)
+        {
+            throw new ArgumentNullException("This user not found");
+        }
+
+        var fileDto = new NewEasySpeakFileDto()
+        {
+            Stream = file.OpenReadStream(),
+            FileName = file.FileName
+        };
+
+        var uploadFileDto = await _fileService.AddFileAsync(fileDto);
+        var profilePhoto = await _context.EasySpeakFiles.FirstOrDefaultAsync(f => f.Id == uploadFileDto.Id);
+
+        if (profilePhoto == null || profilePhoto.Url == null)
+        {
+            throw new ArgumentNullException("This file not found");
+        }
+
+        user.ImageId = uploadFileDto.Id;
+        profilePhoto.UserId = user.Id;
+        await _context.SaveChangesAsync();
+
+        return profilePhoto.Url;
+    }
+
+    private async Task<string> GetProfileImageUrl(long? imageId)
+    {
+        var profileImage = await _context.EasySpeakFiles.FirstOrDefaultAsync(f => f.Id == imageId);
+        return profileImage?.Url ?? "";
+    }
+
+    public async Task<TagDto[]> GetUserTags()
+    {
+        var userId = _authService.UserId;
+        var user = await _context.Users.Include(u => u.Tags).FirstOrDefaultAsync(u => u.Id == userId) ?? throw new ArgumentException($"Failed to find the user with id {userId}");
```

Comment:
```suggestion
        var user = await _context.Users.Include(u => u.Tags).FirstOrDefaultAsync(u => u.Id == userId) 
             ?? throw new ArgumentException($"Failed to find the user with id {userId}");
```

Reply:
You can get tags by
_context.Users.Include(...).FirstOrDefault(...).Select(u => u.Tags)

Not sure if Include will be needed here

Reply:
I have to get the `user` anyway, because I update not only tags but a user as well. I think it's a correct way
` var user = await _context.Users.Include(u => u.Tags).FirstOrDefaultAsync(a => a.Id == userId) ?? throw new ArgumentException($"Failed to find the user with id {userId}");
        _mapper.Map(userDto, user);`

Reply:
In this function you just get tags, you don't make any updates here

Reply:
Updates are done here `_mapper.Map(userDto, user);` or I miss something

Reply:
I don't undersyand what I'm doing wrong, pls explain
`public async Task<UserDto> UpdateUser(UserDto userDto)
    {
        var userId = _authService.UserId;

        // get user from Db with tags
        var user = await _context.Users.Include(u => u.Tags).FirstOrDefaultAsync(a => a.Id == userId) ?? throw new ArgumentException($"Failed to find the user with id {userId}");

        // update user entity with data from request
        _mapper.Map(userDto, user);

        // clear tags
        user.Tags.Clear();

        // set users tags the same as in request
        var dbTags = await _context.Tags.Select(el => el.Name).ToListAsync();
        if (userDto.Tags != null)
        {
            user.Tags = userDto.Tags.Where(el => dbTags.Contains(el.Name)).Select(t => _mapper.Map<TagDto, Tag>(t)).ToList();
        }

        // save changes in db
        await _context.SaveChangesAsync();

        return _mapper.Map<User, UserDto>(user);
    }`

Reply:
![image](https://user-images.githubusercontent.com/57347711/229312721-127f8cf8-b722-4267-8512-43d353a4b925.png)
![image](https://user-images.githubusercontent.com/57347711/229312725-923b2ed0-9ee8-45df-8216-d3a922d19f47.png)
I commented this function, it doesn't have anything connected to updateUser...

Reply:
fixed